## Загрузка и первичный просмотр данных

-Импортирована библиотека pandas.

-Загружен файл salary.csv с разделителем ; в датафрейм date_df.

-Выполнен вывод датафрейма с помощью display(date_df).

-С помощью date_df.info() получена информация о структуре данных: количество строк и столбцов, типы данных, наличие пропусков.

In [2]:
import pandas as pd
date_df = pd.read_csv('salary.csv', sep = ';')
display(date_df)
date_df.info()

work_year employment_type               job_title      salary  \
0       2020.0              FT          Data SCIENTIST     70000.0   
1       2020.0              FT    Product Data Analyst     20000.0   
2       2020.0              FT            Data Analyst     72000.0   
3       2020.0              FT          Data Scientist  11000000.0   
4       2020.0              FT          Data Scientist     45000.0   
..         ...             ...                     ...         ...   
396     2022.0              FT           Data Engineer    126000.0   
397     2022.0              FT            Data Analyst    129000.0   
398     2022.0              FT            Data Analyst    150000.0   
399     2022.0              FT            Data Analyst     67000.0   
400     2022.0              FT  Data Analytics Manager    150260.0   

     salary_in_usd employee_residence company_location company_size  
0          79833.0                 DE               DE            L  
1          20000.0                 HN               HN            S  
2          72000.0                 US               US            L  
3          35735.0                 HU               HU            L  
4          51321.0                 FR               FR            S  
..             ...                ...              ...          ...  
396       126000.0                 US               US            M  
397       129000.0                 US               US            M  
398       150000.0                 US               US            M  
399        67000.0                 CA               CA            M  
400       150260.0                 US               US            M  

[401 rows x 8 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 401 entries, 0 to 400
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   work_year           401 non-null    float64
 1   employment_type     401 non-null    object 
 2   job_title           401 non-null    object 
 3   salary              398 non-null    float64
 4   salary_in_usd       401 non-null    float64
 5   employee_residence  401 non-null    object 
 6   company_location    401 non-null    object 
 7   company_size        401 non-null    object 
dtypes: float64(3), object(5)
memory usage: 25.2+ KB


## Общий анализ
-date_df.head(20) — просмотрены первые 20 строк датафрейма для оценки содержимого.

-date_df.info() — дополнительно проверена структура данных и наличие пустых значений.

-date_df.describe() — проведён описательный статистический анализ (среднее, стандартное отклонение, минимум, максимум, квартили).

-date_df.columns — выведены названия всех столбцов датафрейма.

In [66]:
print(date_df.head(20))
print(date_df.info())
print(date_df.describe())
print(date_df.columns)

    level_0  index  work_year employment_type                 job_title  \
0         0      0     2020.0              FT            Data Scientist   
1         1      1     2020.0              FT      Product Data Analyst   
2         2      2     2020.0              FT              Data Analyst   
3         3      3     2020.0              FT            Data Scientist   
4         4      4     2020.0              FT            Data Scientist   
5         5      5     2020.0              FT            Data Scientist   
6         6      6     2020.0              FT            Data Scientist   
7         7      7     2020.0              FT              Data Analyst   
8         8      8     2020.0              FT              Data Analyst   
9         9      9     2020.0              FT             Data Engineer   
10       10     10     2020.0              FT      Product Data Analyst   
11       11     11     2020.0              FT             Data Engineer   
12       12     12     20

## вывод
#### (дада, я дурдачок и делаю это после всех заданий, извените пожалуйста, обещаю в след раз читать методу внимательно и по порядку Простите `(*>﹏<*)′)
С помощью метода `head(20)` были просмотрены первые строки датасета, чтобы понять структуру таблицы.  

Датасет содержит сведения о зарплатах специалистов в сфере анализа данных. Поля включают: год (`work_year`), тип занятости (`employment_type`), должность (`job_title`), зарплату в валюте компании (`salary`) и в долларах (`salary_in_usd`), страну проживания сотрудника (`employee_residence`), страну офиса компании (`company_location`) и размер компании (`company_size`). 
Метод `info()` показал, что в датасете 346 строк и 10 столбцов. Обнаружены 2 пропущенных значения в поле `salary`. 
Типы данных — числовые (`int64`, `float64`) и категориальные (`object`).  

## Первое задание 
### Групировка и подсчёт количества в стобце локация
#### Описание
Группировка – "`employment_type`" и количество каждой локации "`company_location`". 
Задание означает что нужно «сгрупировать» данные по столбцу "`employment_type`" и подсчитать количество каждой локации. Группировка в pandas осуществляется с помощью `groupby()`:
метод принимает в качестве аргумента название столбца, по которому нужно группировать (например, год, пол, категория). В данном случае `employment_type`.
метод возвращает объект типа DataFrameGroupBy , то есть сгруппированные данные, к которым можно применять различные методы (`count, sum, min, max, mean, last`). 
В методических указаниях есть пример подходящий для данного задания: 
`df.groupby('column1')['n'].count() # группировка по column1 и подсчет количества элементов в конкретном столбце n`
	Соответственно выполняем следующий код на наших данных. 
*Наивно применяю функцию из методы в надежде что это сработает)*

In [3]:
date_df.groupby('employment_type')['company_location'].count() 

employment_type
FL      2
FT    394
PT      5
Name: company_location, dtype: int64

### Вторая попытка
Конечно, этого было недостаточно, т.к. результат не похож на фрагмент, представленный в задание. 
Далее была вторая попытка применить метод `value_counts()`, который анализирует столбец, показывает каждое уникальное значение, прим этом подсчитывая частоту его встречаемости в списке, результат это пары: значение и частота, отсортированные по убыванию.

In [4]:
print(date_df.groupby('employment_type')['company_location'].value_counts())

employment_type  company_location
FL               US                    2
FT               US                  252
                 GB                   37
                 CA                   19
                 IN                   13
                 DE                   12
                 FR                   11
                 ES                    9
                 GR                    9
                 AT                    3
                 MX                    3
                 TR                    3
                 NG                    2
                 PK                    2
                 PL                    2
                 AU                    1
                 BR                    1
                 CH                    1
                 CL                    1
                 HN                    1
                 HU                    1
                 IL                    1
                 IQ                    1
                 IR    

## Вывод по первому заданию
В результате группировки по типу занятости и локации компании видно, что подавляющее большинство работников представлено в категории Full-time (FT). Наибольшее количество записей относится к США (252 случая), что указывает на доминирование американского рынка в наборе данных. Freelance (FL) встречается крайне редко (2 случая в США), а Part-time (PT) также имеет минимальное количество записей, распределённых по разным странам. Это позволяет сделать вывод, что набор данных в основном описывает сотрудников, работающих на полной занятости в крупных компаниях, преимущественно с головными офисами в США.

## Второе задание
Попробуем аналогичную логику первого задания, но переименовать столбец и добавить стортировку
Использован метод groupby для группировки данных по году работы ('work_year') и размеру компании ('company_size').

С помощью 'value_counts()' подсчитано количество записей для каждой комбинации.

Результат преобразован в датафрейм с помощью '.reset_index()'.

Выполнена сортировка по столбцу count в порядке убывания ('ascending = False').

Итоговый датафрейм 'd_q2' выведен на экран ('print(d_q2)'), что позволяет определить наиболее часто встречающиеся комбинации года работы и размера компании.

In [5]:
d_q2 = date_df.groupby('work_year')['company_size'].value_counts().reset_index()
d_q2 = d_q2.sort_values(by='count',ascending = False)
print(d_q2)

   work_year company_size  count
7     2022.0            M    222
4     2021.0            L     57
5     2021.0            M     30
8     2022.0            L     27
0     2020.0            L     21
6     2021.0            S     18
1     2020.0            S     15
2     2020.0            M      5
9     2022.0            S      5
3     2020.0        Large      1


#### Пу пу пу
Тут стало понятно что я немного(много)дурдочка - данные нужно было предобработать... 
Предобработка данных:
- С помощью функции `date_df.duplicated().sum()` было обнаружено 55 явных дубликатов, которые были удалены с помощью `date_df = date_df.drop_duplicates().reset_index(drop=True)`
- Были проанализированные уникальные значения во всех столбцах кроме `salary` и s`alary_in_usd` и переименнованы неверно названные. 
- В столбце salary обнаружены 3 пропущенных значения, тогда как в 'salary_in_usd' отсутствуют пропуски. Так как именно 'salary_in_usd' используется для анализа, отсутствующие значения в 'salary' не влияют на результаты и были оставлены без заполнения.

In [7]:
print(date_df.duplicated().sum()) #выводим количество явных дубликатов
date_df = date_df.drop_duplicates().reset_index(drop=True) #удаляем дубликаты
print(date_df['work_year'].unique()) # Поиск уникальных значений в указаном столбце
print(date_df['company_size'].unique()) # Поиск уникальных значений в указаном столбце
print(date_df['employment_type'].unique()) # Поиск уникальных значений в указаном столбце
print(date_df['job_title'].unique()) # Поиск уникальных значений в указаном столбце
print(date_df['employee_residence'].unique()) # Поиск уникальных значений в указаном столбце
print(date_df['company_location'].unique()) # Поиск уникальных значений в указаном столбце


0
[2020. 2021. 2022.]
['L' 'S' 'M' 'Large']
['FT' 'PT' 'FL']
['Data SCIENTIST' 'Product Data Analyst' 'Data Analyst' 'Data Scientist'
 'Data Engineer' 'Machine Learning Manager' 'Data Analytics Engineer'
 'Data Science Engineer' 'Machine Learning Developer'
 'Data Analytics Manager' 'Head of Data Science'
 'Head of Machine Learning' 'NLP Engineer' 'Data Analytics Lead'
 'DataScientist' 'Data AnalyticsManager']
['DE' 'HN' 'US' 'HU' 'FR' 'IN' 'PK' 'JP' 'GR' 'MX' 'CA' 'AT' 'NG' 'PH'
 'GB' 'ES' 'IT' 'PL' 'BG' 'NL' 'IQ' 'UA' 'SG' 'RU' 'MT' 'CL' 'RO' 'IR'
 'VN' 'BR' 'HK' 'TR' 'RS' 'AR' 'DZ' 'AU' 'CH']
['DE' 'HN' 'US' 'HU' 'FR' 'IN' 'PK' 'JP' 'GR' 'MX' 'CA' 'AT' 'NG' 'GB'
 'ES' 'IT' 'LU' 'PL' 'NL' 'IQ' 'UA' 'IL' 'RU' 'MT' 'CL' 'IR' 'BR' 'VN'
 'TR' 'DZ' 'MY' 'AU' 'CH']


In [8]:
date_df['company_size'] = date_df['company_size'].replace('Large', 'L')

duplicates_job_title = ['Data SCIENTIST', 'DataScientist'] # список неправильных названий
name_job_title = 'Data Scientist'# правильное название
date_df['job_title'] = date_df['job_title'].replace(duplicates_job_title, name_job_title) 
date_df['job_title'] = date_df['job_title'].replace('Data AnalyticsManager', 'Data Analytics Manager')
print(date_df['company_size'].unique()) # Поиск уникальных значений в указаном столбце
print(date_df['job_title'].unique()) # Поиск уникальных значений в указаном столбце


['L' 'S' 'M']
['Data Scientist' 'Product Data Analyst' 'Data Analyst' 'Data Engineer'
 'Machine Learning Manager' 'Data Analytics Engineer'
 'Data Science Engineer' 'Machine Learning Developer'
 'Data Analytics Manager' 'Head of Data Science'
 'Head of Machine Learning' 'NLP Engineer' 'Data Analytics Lead']


In [9]:
d_q2 = date_df.groupby('work_year')['company_size'].value_counts().reset_index()
d_q2 = d_q2.sort_values(by='count',ascending = False)
print(d_q2)

   work_year company_size  count
6     2022.0            M    174
3     2021.0            L     55
4     2021.0            M     30
7     2022.0            L     23
0     2020.0            L     22
5     2021.0            S     18
1     2020.0            S     15
2     2020.0            M      5
8     2022.0            S      4


## Вывод по второму заданию
По результатам группировки можно сделать вывод, что в датасете преобладают компании среднего размера (M), особенно в 2022 году (174 записи). Компании большого размера (L) стабильно занимают значимую долю, а малые компании (S) встречаются значительно реже, причём их количество снижается от 2020 к 2022 году. Таким образом, выборка данных в основном отражает сотрудников средних компаний 2022 и крупных компаний 2021, при этом малые компании представлены минимально.

# Задание 3
Сводная таблица по запрплатам и должностям
С помощью метода `pivot_table()` была построена сводная таблица, где в строках указаны должности, а в значениях — средняя зарплата. С помощью `reset_index()` индекс был преобразован в обычный столбец, чтобы таблица имела привычный вид. 
Методом `round(2)` значения были округлены до двух знаков после запятой. 
Столбец с зарплатами был переименован с 'salary_in_usd' на более понятное название "зарплата" с помощью функции 'rename()'.
Для отображения должностей от самой высокой к самой низкой средней зарплате применена сортировка 'sort_values(by="зарплата", ascending=False)'.

In [10]:
# Сводная таблица: средняя зарплата по должностям
pivot = date_df.pivot_table(index=['job_title'], values='salary_in_usd',aggfunc='mean')
pivot = pivot.reset_index()
pivot["salary_in_usd"] = pivot["salary_in_usd"].round(2)
pivot = pivot.rename(columns={"salary_in_usd": "зарплата"})
pivot = pivot.sort_values(by="зарплата", ascending=False)
display(pivot)

job_title   зарплата
2          Data Analytics Lead  405000.00
7         Head of Data Science  146718.75
3       Data Analytics Manager  127134.29
10    Machine Learning Manager  117104.00
4                Data Engineer  109295.46
6               Data Scientist  102623.46
9   Machine Learning Developer   89395.50
0                 Data Analyst   88597.23
8     Head of Machine Learning   79039.00
5        Data Science Engineer   75803.33
1      Data Analytics Engineer   64799.25
11                NLP Engineer   37236.00
12        Product Data Analyst   13036.00

## Вывод третьему заданию
По результатам анализа видно, что наибольшие средние зарплаты характерны для руководящих должностей. Средний уровень зарплат наблюдается у специальностей, которые входят в число наиболее востребованных специалистов. Наименьшие значения характерны для узкопрофильных позиций, что может быть связано как с меньшим количеством данных по ним, так и с особенностями рынка. В целом, таблица отражает закономерность: чем выше управленческая роль, тем выше средняя зарплата.

## Задание 4
Используя уже использованные ранее методы строим таблицу
`pivot_table()` – построение сводной таблицы: строки = `work_year`, столбцы = `employment_type`, значения = медианная зарплата (`salary_in_usd`). `aggfunc="median"` – расчёт медианного значения вместо среднего.


In [64]:
pivot2 = date_df.pivot_table(values='salary_in_usd', index='work_year', columns='employment_type', aggfunc='median')
pivot2 = pivot2.sort_values(by='work_year',ascending = False)
display(pivot2)

employment_type        FL        FT       PT
work_year                                   
2022.0           100000.0  115967.0  77478.5
2021.0            20000.0   79598.5  40047.5
2020.0                NaN   68428.0  21669.0

## Вывод по четвертому заданию
Медианные зарплаты растут по всем типам занятости с 2020 по 2022 гг. Наибольшие значения наблюдаются у FT, достигнув 116 тыс.  в 2022 году. FL представлен только с 2021 года, при этом медиана резко выросла до 100 тыс. в 2022 году. PT имеет значительно более низкие зарплаты, но также демонстрирует рост.

# ОБЩИЙ ВЫВОД
Анализировался набор данных **salary.csv**, содержащий сведения о зарплатах специалистов в области данных за **2020–2022 годы**: тип занятости, должность, страна проживания, страна офиса, размер компании и доход в долларах США.

### Предобработка данных

* Чтение файла через `pd.read_csv()`.
* Первичный обзор с помощью `head()`, `info()`, `describe()`.
* Удаление **55 явных дубликатов** методом `drop_duplicates()`.
* Выявление и устранение некорректных значений категориальных признаков.
* Обнаружено **3 пропуска в salary** при отсутствии их в `salary_in_usd`; решение — оставить без заполнения, так как для анализа использовался именно `salary_in_usd`.
* Приведение типа `work_year` к **int**.

### Основные результаты анализа

* **Группировка по типу занятости и локации (задание 1):** подавляющее большинство сотрудников работают **Full-time**, основная часть данных относится к **США**; **Part-time** и **Freelance** встречаются редко.
* **Группировка по году и размеру компании:** наибольшая часть данных приходится на **2022 год**; преобладают компании среднего размера (**M**), доля малых компаний (**S**) снижается.
* **Сводная таблица по должности и зарплате в долларах:** руководящие должности имеют наибольшие средние зарплаты, тогда как узкопрофильные роли получают заметно меньше.
* **Сводная таблица по году и типу занятости:** медианные зарплаты стабильно растут с 2020 по 2022 гг.; **FT** выше, чем **PT**, а в 2022 г. **FL** приблизился по медиане к FT.
* **Средняя зарплата** по выборке составила ≈ **105 895 USD**, медианная — более **100 000 USD**; диапазон колеблется от **2 859 до 412 000 USD**.

### Итог

Данные отражают рынок труда специалистов в области данных за **2020–2022 гг.**:

* доминируют контракты **полной занятости**;
* основная часть компаний находится в **США**;
* преобладают **средние и крупные работодатели**;
* уровень зарплат демонстрирует рост по всем категориям занятости.
